controller/share_manager_controller.go

```diff
@@ -1110,19 +1304,103 @@ func (c *ShareManagerController) createPodManifest(sm *longhorn.ShareManager, an
+		return
+	}
+
+	// Consider it stale if there is a lease-holding node, if it has been renewed at least once
+	// since acquisition by that holder, if the duration is a sane value, and if the time of renewal
+	// is longer ago than the lease duration.
+	holder = *lease.Spec.HolderIdentity
+	if holder == "" {
+		return
+	}
+	if lease.Spec.RenewTime == lease.Spec.AcquireTime {
+		log.Warnf("Lease for %v held by %v has never been renewed by share-manager", leaseName, holder)
+		return
+	}
+	if *lease.Spec.LeaseDurationSeconds < shareManagerLeaseDurationSeconds {
+		log.Warnf("Lease for %v has a crazy value for duration: %v seconds.  Ignoring.", leaseName, *lease.Spec.LeaseDurationSeconds)
+		return
+	}
+	expireTime := lease.Spec.RenewTime.Add(time.Duration(*lease.Spec.LeaseDurationSeconds) * time.Second)
+	if time.Now().After(expireTime) {
+		log.Warnf("Lease for %v held by %v is stale, expired %v ago", leaseName, holder, time.Since(expireTime))
+		stale = true
+	}
+
+	return
+}
+
+// isResponsibleFor in most controllers only checks if the node of the current owner is known
+// by kubernetes to be down.  But in the case where the lease is stale or the node is unschedulable
+// we want to transfer ownership and mark the node as delinquent for related status checking.
 func (c *ShareManagerController) isResponsibleFor(sm *longhorn.ShareManager) (bool, error) {
-	// We prefer keeping the owner of the share manager CR to be the same node
-	// where the share manager pod is running on.
+	log := getLoggerForShareManager(c.logger, sm)
+
+	// If the lease is stale, we assume the owning node is down but not officially dead.
+	// Some node has to take over, and it might as well be this one, if another one
+	// has not already.
+	leaseExpired, leaseHolder, err := c.isShareManagerPodStale(sm)
+	if err == nil && leaseExpired {
+		// Avoid race between nodes by checking for an existing interim owner.
+		if leaseHolder != sm.Status.OwnerID {
+			log.Infof("Interim owner %v already took responsibility for stale lease-holder %v", sm.Status.OwnerID, leaseHolder)
+			return false, nil
+		}
+
+		log.Infof("Interim owner %v taking responsibility for stale lease-holder %v", c.controllerID, leaseHolder)
+
+		// Mark expired lease holder as delinquent - best effort.
+		node, err := c.ds.GetNode(leaseHolder)
+		if err != nil {
+			log.WithError(err).Warnf("Failed to get node to set delinquent condition for %v", leaseHolder)
+		} else {
+			node.Status.Conditions = types.SetConditionAndRecord(node.Status.Conditions,
+				longhorn.NodeConditionTypeDelinquent, longhorn.ConditionStatusTrue,
+				string(longhorn.NodeConditionReasonMissedLeaseRenewal),
+				fmt.Sprintf("Node %v marks %v as delinquent", c.controllerID, leaseHolder),
+				c.eventRecorder, node, corev1.EventTypeWarning)
+
+			if _, err := c.ds.UpdateNodeStatus(node); err != nil {
+				log.WithError(err).Warnf("Failed to update node to set delinquent condition for %v", leaseHolder)
+			}
```

Comment:
I think this logic should be moved to that main share-manager sync loop

Reply:
Done.

Reply:
Sorry. Mark as unresolve for now. I might need to rethink about this a bit because there might be potentially a race condition in which share-manager controller repeatedly set node to Delinquent and node controller revert it back to false (in case a false alarm).

Reply:
Also, it leads to repeated delete/create cycles on the share-manager pod, even in the non-false-alarm case.  I reverted it.